test/timeout.bats

```diff
@@ -0,0 +1,252 @@
+#!/usr/bin/env bats
+
+load helpers
+
+function setup() {
+	helptext=$(crictl runp -h)
+	if [[ "$helptext" != *"cancel-timeout"* ]]; then
+		skip "must have a crictl with the -T option to test CRI-O's timeout handling"
+	fi
+	setup_test
+}
+
+function teardown() {
+	cleanup_test
+}
+
+function create_conmon() {
+	timeout=$1
+	cat > "$TESTDIR"/tmp_conmon << EOF
+#!/bin/sh
+if [[ "\$1" != "--version" ]]; then
+	sleep $timeout
+fi
+$CONMON_BINARY \$@
+EOF
+	chmod +x "$TESTDIR/tmp_conmon"
+}
+
+@test "should not clean up pod after timeout" {
+	create_conmon 2s
+	# need infra container so we can timeout in conmon
+	CONTAINER_DROP_INFRA=false CONTAINER_CONMON="$TESTDIR/tmp_conmon" start_crio
+	run crictl runp -T 2s "$TESTDATA"/sandbox_config.json
+	echo "$output"
+	[[ "$output" == *"context deadline exceeded"* ]]
+	[ "$status" -ne 0 ]
+
+	# allow cri-o to catch up
+	sleep 4s
+
+	# cri-o should not report any pods
+	pods=$(crictl pods -q)
+	[[ -z "$pods" ]]
+
+	created_ctr_id=$("$CONTAINER_RUNTIME" --root "$RUNTIME_ROOT" list -q)
+	[[ ! -z "$created_ctr_id" ]]
+
+	run crictl runp "$TESTDATA"/sandbox_config.json
+	echo "$output"
+	[ "$status" -eq 0 ]
+	[[ "$output" == *"$created_ctr_id"* ]]
+}
+
+@test "should not clean up container after timeout" {
+	create_conmon 2s
+	CONTAINER_CONMON="$TESTDIR/tmp_conmon" start_crio
+	pod_id=$(crictl runp "$TESTDATA"/sandbox_config.json)
+
+	run crictl create -T 2s $pod_id "$TESTDATA"/container_config.json "$TESTDATA"/sandbox_config.json
+	echo "$output"
+	[[ "$output" == *"context deadline exceeded"* ]]
+	[ "$status" -ne 0 ]
+
+	# allow cri-o to catch up
+	sleep 4s
+
+	# cri-o should not report any pods
+	ctrs=$(crictl ps -aq)
+	[[ -z "$ctrs" ]]
+
+	# cri-o should have created a container
+	created_ctr_id=$("$CONTAINER_RUNTIME" --root "$RUNTIME_ROOT" list -q | grep -v $pod_id)
+	[[ ! -z "$created_ctr_id" ]]
+
+	run crictl create $pod_id "$TESTDATA"/container_config.json "$TESTDATA"/sandbox_config.json
+	[ "$status" -eq 0 ]
+	[[ "$output" == *"$created_ctr_id"* ]]
+}
+
+@test "should clean up pod after timeout if request changes" {
+	create_conmon 2s
+	# need infra container so we can timeout in conmon
+	CONTAINER_DROP_INFRA=false CONTAINER_CONMON="$TESTDIR/tmp_conmon" start_crio
+	run crictl runp -T 2s "$TESTDATA"/sandbox_config.json
+	echo "$output"
+	[[ "$output" == *"context deadline exceeded"* ]]
+	[ "$status" -ne 0 ]
+
+	# allow cri-o to catch up
+	sleep 4s
+
+	created_ctr_id=$("$CONTAINER_RUNTIME" --root "$RUNTIME_ROOT" list -q)
+	[[ ! -z "$created_ctr_id" ]]
+
+	# we should create a new pod and not reuse the old one
+	run crictl runp <(cat "$TESTDATA"/sandbox_config.json | jq '.metadata.attempt = 2')
```

Comment:
`cat` and the pipe are redunant here, `jq '.metadata.attempt = 2' "$TESTDATA"/sandbox_config.json` would do